### DISCUSSION TITLE
Rules for uncertainty parameters #254

### DISCUSSION THREAD

**Discussion by rtle76** [HydrologicEngineeringCenter/HEC-FDA#262] [General]
[CurrentFDADistributions.xlsx](https://github.com/HydrologicEngineeringCenter/HEC-FDA/files/7984126/CurrentFDADistributions.xlsx)

I've attached a spreadsheet that details how uncertainties are imported into FDA 1.4.3. I referenced the FDA manual and the sensitivity report Bob Carl authored (HEC-FDA Test Triangular Uncertainty about Structure Parameters). Recommendations about how to improve or reformat are welcome. I'm going to run a similar test that Bob ran to see if I can recreate the results. It's not clear how the log-normal distribution is being specified to me.

**Comment by rtle76** [comment]
I was bogged down trying to figure out how Bob went from FFE with a .2 st dev in the normal dist to .000869 in the log normal dist. I kept trying to figure out the math of how he converted from .2 to .000869, but I'm not sure that's what he actually did. I'm under the impression now that you can't convert the parameters of a normal dist to log-normal without the underlying dataset unless you're ok with X's and Y's in the parameters, so he may have just made up the .000869 value

**Reply by rnugent3** [reply]
I think you're right - you can't take the exp of the standard deviation of a log-normal distribution to get the standard deviation of normal. There are ways to approximate. 

The log normal is essentially a normal distribution of logged data.

**Comment by rnugent3** [comment]
I think the document above is a good start. I have some thoughts:

- The most important part of this assignment is to identify what the parameter value should be in 2.0 syntax given the parameter value in 1.4.3 syntax. This means that values should not be left blank - e.g. you cannot construct a normal distribution without the mean. So we need to know what the mean and standard deviation should be for 2.0 normal distribution given the parameter values provided by 1.4.3. You can determine what parameter values you need in 2.0 by reviewing the constructor of each distribution. 
- I would like to be more specific about the units of measurement. This means two things off the top of my head: (1) Let's not talk about standard deviation in percent. Let's talk about coefficient of variation because that is actually what we're dealing with; and (2) if we do have a situation where the units are percent, we should identify percent of what? 
- It would be great to use markdown in this discussion to present the distribution specifications rather than attaching a file.

**Comment by rtle76** [comment]
Thanks for the feedback Richard. I'll spend more time getting a feel for the constructors and working on this tomorrow.

**Comment by rtle76** [comment]
I spent some time thinking about this and working on it. I think I have a good idea about the normal distributions and have some thoughts/questions on the others. 
What exactly do you mean when you say “use markdown in this discussion”? Sorry I’m not quite as experienced with GitHub. I read a bunch about Markdown in Github but not sure what you mean

**Reply by HenryGeorgist** [reply]
the content rendered in these messages is rendered using markdown syntax. So if I wanted to make something italicized i just put the underbar on either side of the word _itlalized_

**Reply by HenryGeorgist** [reply]
you can also write a table using the markdown syntax.

|headername|headername2|
|--|--|
|rowelement|rowelement|

**Reply by HenryGeorgist** [reply]
### Normal Distribution
|parameter|value|
|--|--|
|mean|1|
|standard deviation| coefficent of variation|

**Reply by rnugent3** [reply]
@rtle76 would you put some time on my calendar on Monday? Let's talk through your questions.

**Comment by rtle76** [comment]
The next several posts contain the input/output format for FDA 1.4.3 and the equivalent syntax for FDA 2.0. “STRUCT” indicates the uncertainties will follow for first floor stage, structure value, content-to-structure value ratio, and other-to-structure value ratio in that order. Each of the four uncertainties consist of four columns of values (or blanks) to define the distributions.

**Reply by rtle76** [reply]
Below is the input/output format for first floor stage in FDA 1.4.3 for each of the three distribution types:
| 1.4.3 First Floor Stage | | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **STRUCT** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | STRUCT | N | [Blank] | Std Dev in ft. | [Blank] |
| Triangular | STRUCT | T | [Blank] | Min Value in ft. | Max Value in ft. |
| Log Normal | STRUCT | L | [Blank] | Log Std Dev in ft. | [Blank] |

Below is the equivalent syntax for first floor stage in FDA 2.0 for each of the three distribution types:
| 2.0 First Floor Stage | | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **STRUCT** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | [Begin Uncertainty Syntax] | public Normal() | Mean = 1 | Std Dev in ft. / Sampled FF Stage | [N/A] |
| Triangular | [Begin Uncertainty Syntax] | public Triangular() | Mode = 1 | Min Value in ft. / Sampled FF Stage | Max Value in ft. / Sampled FF Stage |
| Log Normal | [Begin Uncertainty Syntax] | public LogNormal() | Mean = 1 | Log Std Dev in ft. / Log of Sampled FF Stage | [N/A] |

**Reply by rtle76** [reply]
Below is the input/output format for structure value in FDA 1.4.3 for each of the three distribution types:
| 1.4.3 Structure Value | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | N | [Blank] | Std Dev in % of Sampled Structure Value | [Blank] |
| Triangular | T | [Blank] | Min Value in % of Sampled Structure Value | Max Value in % of Sampled Structure Value |
| Log Normal | L | [Blank] | Log Std Dev | [Blank] |

Below is the equivalent syntax for structure value in FDA 2.0 for each of the three distribution types:
| 2.0 Structure Value | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | public Normal() | Mean = 1 | Std Dev in % of Sampled Structure Value / Sampled Structure Value | [N/A] |
| Triangular | public Triangular() | Mode = 1 | Min Value in % of Sampled Structure Value / Sampled Structure Value | Max Value in % of Sampled Structure Value / Sampled Structure Value |
| Log Normal | public LogNormal() | Mean = 1 | Log Std Dev / Log of Sampled Structure Value | [N/A] |

**Reply by rtle76** [reply]
Below is the input/output format for content-to-structure value ratio in FDA 1.4.3 for each of the three distribution types:
| 1.4.3 Content-to-Structure Value Ratio | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | N | Most Likely Value in % | Std Dev in % of Most Likely | [Blank] |
| Triangular | T | Most Likely Value in % | Min Value in % of Most Likely | Max Value in % of Most Likely |
| Log Normal | L | Most Likely Value in % | Log Std Dev | [Blank] |

Below is the equivalent syntax for content-to-structure value ratio in FDA 2.0 for each of the three distribution types. I’m not so sure about these because they’re expressed in percentages, but I took a crack at it:
| 2.0 Content-to-Structure Value Ratio | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | public Normal() | Mean in % | Std Dev in % of Mean | [N/A] |
| Triangular | public Triangular() | Mode in % | Min Value in % of Mode | Max Value in % of Mode |
| Log Normal | public LogNormal() | Log Mean | Log Std Dev | [N/A] |

**Reply by rtle76** [reply]
Below is the input/output format for other-to-structure value ratio in FDA 1.4.3 for each of the three distribution types:
| 1.4.3 Other-to-Structure Value Ratio | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | N | Most Likely Value in % | Std Dev in % of Most Likely | [Blank] |
| Triangular | T | Most Likely Value in % | Min Value in % of Most Likely | Max Value in % of Most Likely |
| Log Normal | L | Most Likely Value in % | Log Std Dev | [Blank] |

Below is the equivalent syntax for other-to-structure value ratio in FDA 2.0 for each of the three distribution types. I’m not so sure about these because they’re expressed in percentages, but I took a crack at it:
| 2.0 Other-to-Structure Value Ratio | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | public Normal() | Mean in % | Std Dev in % of Mean | [N/A] |
| Triangular | public Triangular() | Mode in % | Min Value in % of Mode | Max Value in % of Mode |
| Log Normal | public LogNormal() | Log Mean | Log Std Dev | [N/A] |

**Reply by rnugent3** [reply]
This is really great work, @rtle76. I think that you hit the nail on the head.

**Comment by rtle76** [comment]
Let me know your thoughts on these. 

About the -901 values, I’m not exactly sure where to start. -901 occurs if you leave the uncertainty values blank. Correct me if I’m wrong, but my understanding is we want to be able to take the FDA 1.4.3 export all file and shove it into FDA 2.0 and arrive at the same results. So, we want syntax that will convert -901 into no uncertainty?

**Reply by rnugent3** [reply]
I think that is correct. Where we see -901 the parameter and/or its uncertainty has not been assigned. I think that means that -901 will have to be handled differently for each case. An example would be that we have a content-to-structure value ratio (CSVR) without uncertainty, so there will be -901 for uncertainty value but we have a concrete value for the CSVR, which implies Statistics.Distributions.Deterministic(CSVR). However, if the -901 is in the place of the CSVR, then this parameter is not being used in the study.

**Comment by rnugent3** [comment]
@codymccoy I think that the logic that Jordan laid out here should meet your needs. Please take a look and let us know.

**Reply by HenryGeorgist** [reply]
if there are other -901 values due to the parsing of other parameters lets move that to a different discussion. If the code to parse other data preloads arrays with -901 (which is more of a convention for other programming languages) we can address them separately.

**Comment by rnugent3** [comment]
@rtle76 would you double check the 2.0 tables above? I am pretty sure that triangular distributions of first floor stage and structure value should have a most likely of 100.

**Reply by rtle76** [reply]
It seems like Bob might be suggesting this about first floor elevation in the Structure Parameter Test document. The structure value is 100% of the value indicated in the structure inventory. I would assume this is the case with FFE as well, but I'll do a sensitivity analysis in FDA and have an answer Monday.

**Comment by rnugent3** [comment]
If you look at the table in the document, you’ll see that his investigation applies for all three parameters.

**Comment by rtle76** [comment]
This is the question as I understand it:
	For FFE and structure value, should the most likely value in the triangular distribution be 100 for the code in FDA 2.0?

**Reply by rtle76** [reply]
Structure value: We know the most likely value is left blank for structure value in FDA 1.4.3, according to the user manual (D-11). We know the next two values are the minimum value and the maximum value, and these are global values (ie 90 and 110, not 10 and 10), according to the user manual (D-11) and Bob’s parameter test. Bob lists 100% as the most likely value, which I can only logically conclude is 100% of the structure value listed for each individual structure in the structure inventory. But now I’m starting to think the translation from 1.4.3 is more straightforward because these values are in percentages in 1.4.3:

Below is the equivalent syntax for structure value in FDA 2.0 for each of the three distribution types:
| 2.0 Structure Value | 1st | 2nd | 3rd | 4th |
| ----- | ----- | ----- | ----- | ----- | 
| **Distribution Types** | **Parameter Code** | **Most Likely Value** | **Std Dev OR Min** | **Max** |
| Normal | public Normal() | Mean = 1 | Std Dev in % of Sampled Structure Value / Sampled Structure Value | [N/A] |
| Triangular | public Triangular() | *Mode = 1 X Sampled Structure Value* | *Min Value in % of Sampled Structure Value X Sampled Structure Value* | *Max Value in % of Sampled Structure Value X Sampled Structure Value* |
| Log Normal | public LogNormal() | Mean = 1 | Log Std Dev / Log of Sampled Structure Value | [N/A] |

**Reply by rtle76** [reply]
FFE: In the parameter test report, Bob says, “The definition of the first-floor stage uncertainty is particularly critical because it is not documented that HEC-FDA assumes the most likely value is 100 feet and the triangular distribution is defined globally, i.e. not relatively. For example, the first-floor triangular error might be defined with a minimum value of 98 feet and a maximum value of 103 feet and not with a minimum value of -2 feet and a maximum value of +3 feet. Within the HEC-FDA code, the global values are converted to incremental or relative values for calculations.” 

Again, we know the most likely value is left blank for first floor elevation in FDA 1.4.3, according to the user manual (D-11). We know the next two values are the minimum value and the maximum value, and these are global values (ie 99.8 and 100.2, not .2 and .2), according to the user manual (D-11) and Bob’s parameter test. 

Bob lists 100 feet as the most likely value. Table 9.1 in the user manual (9-11) prescribes uncertainty values based on the measurement method used. I’m not sure, but the only thing I can think is that the sampled value for first floor elevation from the structure inventory is scaled down to 100 ft for calculations and then scaled back up when reporting results (which I think would effectively mean the “100” value is actually percent, not feet as is suggested in the manual and Bob’s report). But that leads to the question: if this isn’t made explicit in the user manual, would the typical user know to adjust their inputs assuming 100ft is the most likely value (and thus the uncertainty parameters should be in percentages)? I would assume not because the user manual and model interface suggest otherwise. 

Do we have the files Bob used for the parameter test report? I’m unable to duplicate his results, in part because he doesn’t give enough information about the other variables required for input into 1.4.3. Having the project files he used would be helpful to understand this.

**Reply by rnugent3** [reply]
I think you have collected enough information and dug into Bob's documentation to confirm that triangular has a most likely value of 100 for structure value and first floor elevation. We should use this value, independent of how the teams may have interpreted the proper use of the triangular distribution. I think you're right - that there is the potential that teams used triangular differently - but I think that misapplication of the triangular distribution is a problem that we can't fix upon importing 1.4.3 study data into 2.0. 

That said, if teams misapplied the triangular distribution and attempt to non-sensical distributions, we should hit an exception, requiring that the problem be addressed.

**Reply by rnugent3** [reply]
Ah - and one more thing. You have mean/most likely values as 1, which is statistically correct, but FDA expects values to be in percent (%) and not decimal form. So for triangular distribution we have a most likely of 100 (%) for FFE and 100 (%) for structure value.
